components/calendar/src/any_calendar.rs

```diff
@@ -28,211 +25,402 @@ define_preferences!(
+                    if let $any_calendar_ident::$variant(c) = value {
+                        Ok(c)
+                    } else {
+                        Err(value)
+                    }
+                }
+            }
+
+            impl<'a> TryFrom<&'a $any_calendar_ident> for &'a $ty {
+                type Error = &'a $any_calendar_ident;
+                fn try_from(value: &'a $any_calendar_ident) -> Result<Self, Self::Error> {
+                    if let $any_calendar_ident::$variant(ref c) = value {
+                        Ok(c)
+                    } else {
+                        Err(value)
+                    }
+                }
+            }
+        )+
+    };
+}
+
+make_any_calendar!(
+    /// This is a calendar that encompasses a selection of calendars from this crate.
+    ///
+    /// This allows for the construction of [`Date`] objects that have their calendar known at runtime.
+    ///
+    /// This can be constructed by calling `.into()` on a concrete calendar type if the calendar type is known at
+    /// compile time. When the type is known at runtime, the [`AnyCalendar::new()`] and sibling methods may be used.
+    ///
+    /// [`Date`] can also be converted to [`AnyCalendar`]-compatible ones
+    /// via [`Date::to_any()`](crate::Date::to_any()).
+    ///
+    /// There are many ways of constructing an AnyCalendar'd date:
+    /// ```
+    /// use icu::calendar::{AnyCalendar, AnyCalendarKind, Date, cal::{Japanese, Gregorian}, types::MonthCode};
+    /// use icu::locale::locale;
+    /// use tinystr::tinystr;
+    /// # use std::rc::Rc;
+    ///
+    /// let locale = locale!("en-u-ca-japanese"); // English with the Japanese calendar
+    ///
+    /// let calendar = AnyCalendar::new(AnyCalendarKind::new(locale.into()));
+    ///
+    /// // This is a Date<AnyCalendar>
+    /// let any_japanese_date = Date::try_new_gregorian(2020, 9, 1)
+    ///     .expect("Failed to construct Gregorian Date.")
+    ///     .to_calendar(calendar)
+    ///     .to_any();
+    ///
+    /// // Construct a date in the appropriate typed calendar and convert
+    /// let japanese_calendar = Japanese::new();
+    /// let japanese_date = Date::try_new_japanese_with_calendar("reiwa", 2, 9, 1,
+    ///                                                         japanese_calendar).unwrap();
+    /// assert_eq!(japanese_date.to_any(), any_japanese_date);
+    ///
+    /// // this is also Date<AnyCalendar>, but it uses a different calendar
+    /// let any_gregorian_date = any_japanese_date.to_calendar(Gregorian).to_any();
+    ///
+    /// // Date<AnyCalendar> does not have a total order
+    /// assert!(any_gregorian_date <= any_gregorian_date);
+    /// assert!(any_japanese_date <= any_japanese_date);
+    /// assert!(!(any_gregorian_date <= any_japanese_date) && !(any_japanese_date <= any_gregorian_date));
+    /// ```
+    #[non_exhaustive]
+    AnyCalendar,
+
+    #[non_exhaustive]
+    AnyDateInner,
+    #[cfg(feature = "unstable")],
```

Comment:
nit: I don't think this should be an input to the macro.

Reply:
how would you make this work cross-crate? require the `unstable` feature to use the macro?

Reply:
I'm not yet convinced I want this to work cross crate. I would like to land this change without having to think about that broader issue, or having public API changes to consider.